File: C++/problem1029.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
#include <cmath>
#include <ctime>
#include <cstdlib>
#include <windows.h>
#include <string>
#include <numeric>

using namespace std;

class Solution {
public:
    int twoCitySchedCost(vector<vector<int>>& costs) {
        sort(costs.begin(), costs.end(), 
                [](vector<int> o1, vector<int> o2) //这像一个匿名函数 没有函数名和返回类型
                {
                    return (o1[0]-o1[1] < o2[0]-o2[1]);                    
                }
        );
        int n = costs.size()/2; //两个整型自动就是整除，不需要强制类型转换 int n = (int)costs.size()/2;
        // int n = costs.size()/2.0; //可能有隐式转换
        int ans = 0;
        for (int k = 0; k < n; ++k)
        {
            ans += costs[k][0] + costs[k+n][1];
        }
        return ans;
    }
};

int main()
{
    //costs中的临时向量可能不需要名称
    vector<vector<int>> costs {vector<int> {10, 20}, vector<int> {30, 200}, vector<int> {400, 50}, vector<int> {30, 20}};
    Solution solu;
    int ans = solu.twoCitySchedCost(costs);
    cout << ans << endl;
    return 0;
}

File: Python/problem1029.py
from typing import List

class Solution:
    def twoCitySchedCost(self, costs: List[List[int]]) -> int:
        # costs.sort(reverse = True, key = lambda x: abs(x[0]-x[1]))
        
        # A, B = [], []
        # d = {}
        # for cost in costs:
        #     if cost[0] < cost[1]:
        #         A.append(cost)
        #     else:
        #         B.append(cost)
        #     if str(cost) in d:
        #         d[str(cost)] += 1
        #     else:
        #         d[str(cost)] = 1
        # A.sort(reverse = True, key = lambda x: x[0])
        # B.sort(reverse = True, key = lambda x: x[1])

        # ans = 0
        # for cost in costs:
        #     if cost[0] < cost[1]:
        #         ans += cost[0]
        #         d[str(cost)] -= 1
        #         mark = 0
        #         while B:
        #             tmp = B.pop()
        #             if d[str(tmp)] > 0:
        #                 ans += tmp[1]
        #                 d[str(tmp)] -= 1
        #                 mark = 1
        #                 break
        #         if mark == 0:
        #             while A:
        #                 tmp = A.pop()
        #                 if d[str(tmp)] > 0:
        #                     ans += tmp[1]
        #                     d[str(tmp)] -= 1
        #                     mark = 1
        #                     break
        #         if mark == 0:
        #             return ans
        #     else:
        #         ans += cost[1]
        #         d[str(cost)] -= 1
        #         mark = 0
        #         while A:
        #             tmp = A.pop()
        #             if d[str(tmp)] > 0:
        #                 ans += tmp[0]
        #                 d[str(tmp)] -= 1
        #                 mark = 1
        #                 break
        #         if mark == 0:
        #             while B:
        #                 tmp = B.pop()
        #                 if d[str(tmp)] > 0:
        #                     ans += tmp[0]
        #                     d[str(tmp)] -= 1
        #                     mark = 1
        #                     break
        #         if mark == 0:
        #             return ans

        costs.sort(key = lambda x: x[0] - x[1]) #x[0] - x[1]可以理解为去A的有利程度
        ans = 0
        n = len(costs) // 2
        # for k in range(0, n):
        #     ans += costs[k][0] + costs[k+n][1]
        # return ans
        for k in range(0, n):
            ans += costs[k][0]
        for k in range(n, 2*n):
            ans += costs[k][1]
        return ans

solu = Solution()
costs = [[10,20],[30,200],[400,50],[30,20]]
print(solu.twoCitySchedCost(costs))